tools/lint-struct-tags/main.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+func main() {
+	if len(os.Args) != 2 {
+		fmt.Fprintf(os.Stderr, "Usage: %s <directory>\n", os.Args[0])
+		os.Exit(1)
+	}
+	dir := os.Args[1]
+	duplicatesFound := false
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() && strings.HasSuffix(path, ".go") && !strings.HasSuffix(path, ".pb.go") &&
+			!strings.HasSuffix(path, "_test.go") {
+			dupes, err := analyzeFile(path)
+			if err != nil {
+				return err
+			}
+			if dupes {
+				duplicatesFound = true
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
+	}
+	if duplicatesFound {
+		os.Exit(1)
+	}
+}
+
+func analyzeFile(file string) (bool, error) {
+	fset := token.NewFileSet()
+	node, err := parser.ParseFile(fset, file, nil, parser.ParseComments)
+	duplicatesFound := false
+	if err != nil {
+		return false, err
+	}
```

Comment:
**[suggestion, non-blocking]**

```suggestion
	if err != nil {
		return false, err
	}
	duplicatesFound := false
```

This is my bad — I think it’s idiomatic to have the `if err != nil {` right next to the line that produces the error.